onnxruntime/core/session/custom_ops.cc

```diff
@@ -163,22 +163,35 @@ common::Status CreateCustomRegistry(gsl::span<OrtCustomOpDomain* const> op_domai
       }
     }
 
+    constexpr uint32_t min_ort_ver_io_opt = 8;   // Minimum ORT version supporting optional input/output.
+    constexpr uint32_t min_ort_ver_io_var = 14;  // Minimum ORT version supporting variadic input/output.
+
     std::vector<ONNX_NAMESPACE::OpSchema> schemas_list;
     for (const auto* op : domain->custom_ops_) {
       ONNX_NAMESPACE::OpSchema schema(op->GetName(op), "custom op registered at runtime", 0);
 
       size_t type_id_counter = 0;
-      auto input_count = op->GetInputTypeCount(op);
+      const size_t input_count = op->GetInputTypeCount(op);
       for (size_t i = 0; i < input_count; i++) {
         onnx::OpSchema::FormalParameterOption option = onnx::OpSchema::FormalParameterOption::Single;
 
-        // Only since the ORT API version 8 and onwards does the OrtCustomOp interface have the relevant methods exposed to query
-        // if an input/output is required/optional. So, query the relevant methods ONLY from API version 8 onwards.
-        if (op->version >= 8 && op->GetInputCharacteristic(op, i) == OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_OPTIONAL) {
-          option = onnx::OpSchema::FormalParameterOption::Optional;
+        // The OrtCustomOp interface did not support the methods to query input/output characteristics before
+        // ORT API version 8. So, query the relevant methods ONLY from API version 8 onwards.
+        if (op->version >= min_ort_ver_io_opt) {
+          const auto characteristic = op->GetInputCharacteristic(op, i);
+
+          // Support for optional and variadic inputs/output was added in versions 8 and 14, respectively.
+          if (characteristic == OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_OPTIONAL) {
+            option = onnx::OpSchema::FormalParameterOption::Optional;
+          } else if ((op->version >= min_ort_ver_io_var) &&
+                     (characteristic == OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_VARIADIC)) {
+            // Only the last input can be variadic.
+            ORT_ENFORCE(i == input_count - 1, "Only the last input to a custom op may be marked variadic.");
```

Comment:
This is not strictly necessary because onnx validation will throw an exception if we don't bail out. However, the onnx exception message may not be as clear/explicit.

Here's what ONNX would throw:
```
Schema error: ONNX Schema VariadicNode: failed validating the check: (inputs_.size() - 1) == i
```